pkg/driver/controller.go

```diff
@@ -0,0 +1,311 @@
+	}
+
+	var (
+		gidMin           int
+		gidMax           int
+		basePath         string
+		provisioningMode string
+		volContext       = make(map[string]string)
+		err              error
+	)
+
+	//Parse parameters
+	volumeParams := req.GetParameters()
+	if value, ok := volumeParams[ProvisioningMode]; ok {
+		provisioningMode = value
+		//TODO: Add FS provisioning mode check when implemented
+		if provisioningMode != "efs-ap" {
+			errStr := "Provisioning mode " + provisioningMode + " is not supported. Only Access point provisioning: 'efs-ap' is supported"
+			return nil, status.Error(codes.InvalidArgument, errStr)
+		}
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", ProvisioningMode)
+	}
+
+	accessPointsOptions := &cloud.AccessPointOptions{
+		CapacityGiB: volSize,
+	}
+
+	if value, ok := volumeParams[FsId]; ok {
+		klog.V(5).Infof("FS ID: %v", value)
+		accessPointsOptions.FileSystemId = value
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", FsId)
+	}
+
+	if value, ok := volumeParams[GidMin]; ok {
+		gidMin, err = strconv.Atoi(value)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "Failed to parse invalid %v: %v", GidMin, err)
+		}
+		if gidMin <= 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "%v must be greater than 0", GidMin)
+		}
+	}
+
+	if value, ok := volumeParams[GidMax]; ok {
+		// Ensure GID min is provided with GID max
+		if gidMin == 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", GidMin)
+		}
+		gidMax, err = strconv.Atoi(value)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "Failed to parse invalid %v: %v", GidMax, err)
+		}
+		if gidMax <= gidMin {
+			return nil, status.Errorf(codes.InvalidArgument, "%v must be greater than %v", GidMax, GidMin)
+		}
+	} else {
+		// Ensure GID max is provided with GID min
+		if gidMin != 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", GidMin)
+		}
+	}
+
+	if value, ok := volumeParams[DirectoryPerms]; ok {
+		accessPointsOptions.DirectoryPerms = value
+	}
+
+	if value, ok := volumeParams[DeleteContent]; ok {
+		accessPointsOptions.DeleteContent, _ = strconv.ParseBool(value)
+	}
+
+	if value, ok := volumeParams[BasePath]; ok {
+		basePath = value
+	}
+
+	// Assign default GID ranges if not provided
+	if gidMin == 0 && gidMax == 0 {
+		gidMin = DefaultGidMin
+		gidMax = DefaultGidMax
+	}
+
+	// Check if file system exists. Describe FS handles appropriate error codes
+	if _, err = d.cloud.DescribeFileSystem(ctx, accessPointsOptions.FileSystemId); err != nil {
+		if err == cloud.ErrNotFound {
+			return nil, status.Errorf(codes.InvalidArgument, "File System does not exist: %v", err)
+		}
+		return nil, status.Errorf(codes.Internal, "Failed to fetch File System info: %v", err)
+	}
+	gid, err := d.gidAllocator.getNextGid(accessPointsOptions.FileSystemId, gidMin, gidMax)
```

Comment:
Say the directory permission is 770 or 740, how will the user using the PV discover the GID and ensure their Pod has it? The old provisioner put a special GID annotation on all PVs https://github.com/kubernetes-retired/external-storage/blob/master/aws/efs/cmd/efs-provisioner/efs-provisioner.go#L190. This PV annotation is basically a field in kubernetes that means: add this group to the consuming Pod's supplemental groups. (Consuming pod meaning the Pod in relation Pod->PVC->PV). It doesn't seem possible in CSI to make this annotation show up on the final PV without updating https://github.com/kubernetes-csi/external-provisioner to do it somehow. I am not sure the usefulness of assigning a random GID without also setting the spcial annotation

Reply:
I ran some tests to see how K8s and the pods behave with different directory permissions. By default, it looks like pods get root(owner) permissions and pods will be able to access the PV.

The random GIDs help with restricting access on EFS. Each GID will be used to create a root directory for access points and the user using a PV will not be able to access other root directories behind different access points which serve other PVs. Also GIDs here serve as unique `ID` for an access point root directory. This prevents access points from re-using existing root directories and prevents data from being due to overwritten/corruption.

Reply:
The UID/GID a pod gets is dependent on the image, PSP enforced on the pod, and pod's security context. There is no guarantee that a pod will get root, this assumption is wrong.

I do not think there is any risk of users accessing directories other than what is under their EFS access point nor do I think GID is a good mechanism for controlling access. Besides, if you are going with the assumption that all pods run as root, and efs does not squash root, then it doesn't matter what GID the directories are. Anyway, all directories provisioned by the driver should be at the same level in the tree, there should never be one directory a child of another.

For a unique ID, I would prefer something truly unique and more useful like PVC UID

Reply:
If I am reading https://docs.aws.amazon.com/efs/latest/ug/accessing-fs-nfs-permissions.html#accessing-fs-nfs-permissions-access-points correctly, the access point will be readable no matter what the uid/gid/psp/etc. is since "When user enforcement is enabled, Amazon EFS replaces the NFS client's user and group IDs with the identity configured on the access point for all file system operations."

Can you test that if the Pod does not have root uid (using one of the ways i mentioned, securitycontext is easiest) that it can still read/write the access point no matter what the access point's uid/gid is

Reply:
To provide additional context, I configured the pod's security context to runAsUser and runAsGroup to be different from that of the GID of the root directory with permissions 700, the pod was able to access the PV. May be it the UID/GID is a determined by a combination of all image,  PSP and security context? - I'll do additional research on this.

Posix UID,GID and directory perms are required parameters to create an [access point root directory](https://docs.aws.amazon.com/efs/latest/ug/efs-access-points.html#enforce-root-directory-access-point).
The driver ensures all root directories created for access points will be under the same base path and one will not be a child of another.

In any case, your concern is valid and currently there doesn't seem to be a way to propagate the GIDs to pods from CSI driver.

Unique ID is a combination of a prefix(efs-csi-ap), GID & a UUID. PVC UID is used as the creation token for access point. It can also be used as the directory name, however, felt the former would be a good choice to identify the directories created for access points.